hacks/looper.scd

```diff
@@ -0,0 +1,89 @@
+/*
+fooloop is inspired by chunk-recorder.scd
+This looper records from SoundIn and writes/replaces buffers directly to the SuperDirt soundLibrary.
+
+Copyright (C) 2019  Michael Hauck (f00b455)
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+How to use in Tidal:
+
+-- record and replace round robin
+d1 $ sound "fooloopRec"
+
+-- stop recording and replacing
+d1 $ silence
+
+-- play recorded loop 3 (bufnamePrefix++counter)
+d5 $ sound "fooloop3" # gain "0.7"
+
+-- reset buffers. But do only one time (d1 $ silence),
+-- since d1 $ once $ sound "fooloopReset" does not work at my installation (1.4.3)
+d1 $ sound "fooloopReset"
+
+*/
+(
+var functions = ();  // make a dictionary of functions
+var recSynth, recBufs, counter = 0, recording = false;
+var whichOrbit = ~dirt.orbits[0];
+var maxTime = 24; // allow a maximum of four seconds, adjust to your needs
+var lat = s.latency + 0.032; // finetune
+var numBuffers = 8; // number of buffers
+var bufnamePrefix = "fooloop"; // soundname prefix
+
+~recBufs.do(_.free); // free them if they are left over (this is why we keep it in the environment variable)
+~recBufs = recBufs = { Buffer.alloc(~dirt.server, ~dirt.server.sampleRate * maxTime, ~dirt.numChannels) } ! numBuffers;
+
+// recorder writes audio from SoundIn to a bufnum
+SynthDef(\fooRecord, { |bufnum|
+	var in = SoundIn.ar([0,1]);
+	RecordBuf.ar(in, bufnum, loop:0, doneAction:2);
+}).add;
+
+// start recording round robin to buffers
+functions[\fooloopRec] = {
+	if(recording) { functions[\fooloopAdd].value };
+	counter = counter + 1;
+	if(counter == numBuffers, {counter = 0});
```

Comment:
`counter` is used as an index into `recBufs`. But the last possible index in `recBufs` is `numBuffers - 1`. You can either skip this lile completely, because you are using `@@` below (which is `wrapAt`), and the counter can just go up forever. Or you can check for  e.g.`counter >= recBufs.lastIndex` and then just use `@` or `at` below.

Reply:
thanks, i will fix it.

Reply:
`counter` is used in `fooloopAdd` to generate the name for the soundLibrary. So it has to be resetted. I changed the `@@` to `@` and updated the condition for the reset.